pkg/kubelet/cm/deviceplugin/manager.go

```diff
@@ -610,58 +610,95 @@ func (m *ManagerImpl) allocateContainerResources(pod *v1.Pod, container *v1.Cont
 		if allocDevices == nil || len(allocDevices) <= 0 {
 			continue
 		}
-		startRPCTime := time.Now()
-		// devicePluginManager.Allocate involves RPC calls to device plugin, which
-		// could be heavy-weight. Therefore we want to perform this operation outside
-		// mutex lock. Note if Allocate call fails, we may leave container resources
-		// partially allocated for the failed container. We rely on updateAllocatedDevices()
-		// to garbage collect these resources later. Another side effect is that if
-		// we have X resource A and Y resource B in total, and two containers, container1
-		// and container2 both require X resource A and Y resource B. Both allocation
-		// requests may fail if we serve them in mixed order.
-		// TODO: may revisit this part later if we see inefficient resource allocation
-		// in real use as the result of this. Should also consider to parallize device
-		// plugin Allocate grpc calls if it becomes common that a container may require
-		// resources from multiple device plugins.
+		// Update internal cached podDevices state.
 		m.mutex.Lock()
-		e, ok := m.endpoints[resource]
+		m.podDevices.insert(podUID, contName, resource, allocDevices, nil)
 		m.mutex.Unlock()
-		if !ok {
-			m.mutex.Lock()
-			m.allocatedDevices = m.podDevices.devices()
+	}
+	return nil
+}
+
+// GetDeviceRunContainerOptions checks whether we have cached containerDevices
+// for the passed-in <pod, container> and returns its DeviceRunContainerOptions
+// for the found one. An empty struct is returned in case no cached state is found.
+func (m *ManagerImpl) GetDeviceRunContainerOptions(pod *v1.Pod, container *v1.Container) *DeviceRunContainerOptions {
+	podUID := string(pod.UID)
+	contName := container.Name
+	invokeRPCBoolFlag := false
+	for k, v := range container.Resources.Limits {
+		resource := string(k)
+		needed := int(v.Value())
+		glog.V(3).Infof("needs %d %s", needed, resource)
+		_, registeredResource := m.healthyDevices[resource]
+		_, allocatedResource := m.allocatedDevices[resource]
+		// Continues if this is neither an active device plugin resource nor
+		// a resource we have previously allocated.
+		if !registeredResource && !allocatedResource {
+			continue
+		}
+		m.mutex.Lock()
+		devices, resp := m.podDevices.containerDevices(podUID, contName, string(k))
```

Comment:
We probably don't need the resp here

Reply:
Do you mean not needed because we would make Allocate each time? 
If yes, please share your opinion on making this a configurable option as i explained in another comment.